rfcs/text/0032-structured-benchmark-programs.md

```diff
@@ -0,0 +1,192 @@
+
+- *Grover's Search with Weak Measurement*: Measurements on an ancilla qubit are utilized to determine whether to continue with another iteration of Grover's search or to stop as amplitude amplification has succeeded. ([Andrés-Martínez et al.](https://iopscience.iop.org/article/10.1088/2058-9565/ac47f1/meta))
+- *Repeat-Until-Success*: Specific operations are applied repeatedly until a desired measurement outcome is achieved, indicating that a desired complex operation has been successfully implemented.
+- *Quantum Metropolis Sampling*: [Temme et al., 2011](https://arxiv.org/abs/0911.3635)
+- *ML-QAE*: [Suzuki et al., 2020](https://arxiv.org/abs/1904.10246)
+
+## Conditionals
+
+*This class of benchmark programs utilizes conditionals (such as `if`) to implement protocols or algorithms.*
+
+- *Quantum Teleportation*: Quantum teleportation is a canonical primitive for quantum communication and distributed computing. It utilizes conditionals based on measurement results to apply the appropriate correction operations. ([Bennett et al., 1993](https://doi.org/10.1103/PhysRevLett.70.1895))
+- *Block Encoding*: Block encoding embeds classical data or operators into quantum states. It underpins advanced algorithms like QSVT. It utilizes conditionals to apply different operations. ([Quantum Singular Value Transformation (QSVT)](https://arxiv.org/abs/1806.01838), [Low & Chuang, 2016](https://arxiv.org/abs/1606.02685))
+
+## Mixed Control Flow
+
+*This class of benchmark programs combines loops and conditionals to implement more complex algorithms or protocols.*
+
+- *Quantum Multiplexers*: Quantum multiplexers apply different unitary operations based on the state of control qubits, effectively implementing conditionals in quantum circuits. ([Quantum multiplexers in circuit synthesis](https://arxiv.org/abs/quant-ph/0410066))
+
+## Dynamic Qubit (Re-)Allocation
+
+*This class of benchmark programs allocates qubits dynamically during execution, based on runtime conditions or within loop bodies.*
+
+- *Iterative Quantum Fourier Transform (iQFT)*: Rather than allocating all qubits upfront, just a single qubit is allocated at the start of execution and then reset during each loop iteration of QFT.
+- *Iterative Quantum Phase Estimation (iQPE)*: Similar to iQFT, iQPE allocates a single qubit (in addition to the phase qubit "psi") and reuses it across iterations to estimate the phase of a unitary operator.
+- *Shor's Algorithm*: The primitives used in Shor's algorithm can similarly be implemented using dynamic qubit reallocation to minimize the total number of qubits required.
+- *Toffoli-heavy Circuits*: Toffoli gates in quantum circuits can be implemented more efficiently using ancilla qubits. Circuits using a larger number of Toffoli gates, such as many arithmetic circuits, can benefit from dynamic qubit allocation to allocate and deallocate ancilla qubits repeatedly inside loop bodies.
```

Comment:
Another example could be the parallelization technique applied in [Quantum Fan-out is Powerful](https://www.theoryofcomputing.org/articles/v001a005/v001a005.pdf). See Figs. 3 and 4.

Reply:
Also added to table; can resolve this and decide above whether it's worth including.